src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
+                output_buffer.as_mut_slice(),
+            )
+            .map_err(|err| {
+                anyhow::anyhow!(format!("error reading output buf size: {:?}", err))
+            })?;
+        let output: Value =
+            rmp_serde::from_slice(output_buffer.as_ref()).map_err(|err| {
+                anyhow::anyhow!(format!("error decoding output buf: {:?}", err))
+            })?;
+        // return the entire size of the output buffer (including i32 size prefix) so it can be passed to dealloc() later
+        let result: (Value, i32) = (output, (size + SIZE_BYTE_COUNT).try_into().unwrap());
+        Ok(result)
+    }
+
+    fn write_udf_input(&mut self, input: &Value) -> anyhow::Result<(i32, i32)> {
+        // serialize input using MessagePack
+        let mut udf_input_buf: Vec<u8> = vec![];
+        input
+            .serialize(&mut Serializer::new(&mut udf_input_buf))
+            .map_err(|err| {
+                anyhow::anyhow!(format!("Error serializing input {:?}", err))
+            })?;
+        // Total input size will be serialized messagepack bytes prepended by the
+        // size of the serialized input (one i32 == 4 bytes)
+        let size_len = std::mem::size_of::<i32>();
+        let udf_input_size: usize = udf_input_buf.len() + size_len;
+        // allocate WASM memory for input buffer
+        // TODO: handle case when allocation fails
+        let udf_input_ptr = self
+            .alloc
+            .call(&mut self.store, udf_input_size.try_into().unwrap())?;
+        let ptr: usize = udf_input_ptr.try_into().unwrap();
+        // write size of input buffer first
+        self.memory
+            .write(&mut self.store, ptr, &udf_input_buf.len().to_ne_bytes())
+            .map_err(|err| {
+                anyhow::anyhow!(format!("Error copying UDF input: {:?}", err))
+            })?;
+        // copy input buffer
+        self.memory
+            .write(&mut self.store, ptr + size_len, udf_input_buf.as_ref())
+            .map_err(|err| {
+                anyhow::anyhow!(format!("Error copying UDF input: {:?}", err))
+            })?;
+        // return the entire size of the output buffer (including i32 size prefix) so it can be passed to dealloc() later
+        Ok((udf_input_ptr, udf_input_size.try_into().unwrap()))
+    }
+
+    pub fn call(&mut self, input: Vec<Value>) -> anyhow::Result<Value> {
+        let args_array = Value::Array(input);
+        let (udf_input_ptr, input_size) = self.write_udf_input(&args_array)?;
+        // invoke UDF
+        let udf_output_ptr = self.udf.call(&mut self.store, udf_input_ptr)?;
+        let (output, output_size) = self.read_udf_output(udf_output_ptr)?;
+        // deallocate both input and output buffers
+        self.dealloc
+            .call(&mut self.store, (udf_input_ptr, input_size))?;
+        self.dealloc
+            .call(&mut self.store, (udf_output_ptr, output_size))?;
+        Ok(output)
+    }
+}
+
+fn get_arrow_value<T>(
```

Comment:
It's kind of strange that this function isn't provided by Arrow/DataFusion to begin with, it feels pretty general. Are you sure it isn't defined in a public module somewhere?

Reply:
Totally agree; I don't know enought about Arrow/DataFusion to be sure, but I couldn't find it anywhere.